### PR TITLE
[microNPU] Refactor base address determination to codegen

### PR DESCRIPTION
**Pull request by manupa-arm**
This commit introduces BaseAddress ObjectRef to determine
base addresses in the codegen for microNPU. This is
required when multiple memory pools become available. Thus,
base addresses could not be statically determined in the
source module.

### CODE REVIEW

**Comment by manupa-arm** [comment]
cc : @grant-arm @Mousius @ekalda @mbaret

**Comment by manupa-arm** [comment]
a friendly ping for reviews! TIA.

File: python/tvm/relay/backend/contrib/ethosu/tir_to_cs_translator.py [hunk 1]
```diff
@@ -36,15 +36,15 @@ class BufferType(Enum):
 
     constant = auto()
     input_or_output = auto()
-    scratch = auto()
+    runtime_allocate = auto()
```
Comment (mbaret) [inline python/tvm/relay/backend/contrib/ethosu/tir_to_cs_translator.py:39]: I'm not sure if 'runtime_allocate'/'runtime_allocation' is the best name for this as it mixes up purpose and allocation strategy. I think we can stick to 'scratch' here and rely on the 'is_runtime_allocation' field on BaseAddress to provide the further info.

File: src/relay/backend/contrib/ethosu/utils.h [hunk 2]
```diff
@@ -34,47 +34,90 @@ namespace relay {
 namespace contrib {
 namespace ethosu {
 
+/*!
+ * \brief Base addresses are input pointers to
+ * the driver that get accessed by produced
```
Comment (mbaret) [inline src/relay/backend/contrib/ethosu/utils.h:39]: incomplete?

**Comment by manupa-arm** [comment]
Thanks @mbaret for the quick review! and I believe I've addressed the comments.

**Comment by manupa-arm** [comment]
a friendly ping for @grant-arm @Mousius

**Review by Mousius — APPROVED** [review]
This looks good to me @manupa-arm, I'd be cautious of over-using `utils` as a general place to put things though as it's unclear where the functional boundary is. I'd suggest renaming it to `objects.cc` which is one step less vague, or splitting it further in future.

**Comment by grant-arm** [comment]
Thanks for this @manupa-arm , it looks good to me.

**Comment by manupa-arm** [comment]
@Mousius that make sense. lets rename it to be objects in a follow up. This passed CI :) finally. Can we take this in ?

**Pull request merged**